keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -0,0 +1,306 @@
+---
+title: Defaulting for Custom Resources
+authors:
+  - "@sttts"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@lavalamp"
+  - "@liggitt"
+  - "@mbohlool"
+  - "@apelisse"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+editor:
+  name: "@sttts"
+creation-date: 2019-04-26
+last-updated: 2019-04-26
+status: implementable
+see-also:
+ - "/keps/sig-api-machinery/20180731-crd-pruning.md"
+ - "/keps/sig-api-machinery/20190425-structural-openapi.md"
+---
+
+# Defaulting for Custom Resources
+
+## Table of Contents
+
+   * [Defaulting for Custom Resources](#defaulting-for-custom-resources)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Examples](#examples)
+      * [References](#references)
+         * [Test Plan](#test-plan)
+         * [Graduation Criteria](#graduation-criteria)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+Defaulting is a fundamental step in the processing of API objects in the request pipeline of the kube-apiserver. Defaulting happens during deserialization, i.e. after decoding of a versioned object, **but before** conversion to a hub type.
+
+Defaulting is implemented for most native Kubernetes API types and plays a crucial role for API compatibility when adding new fields. CustomResources do not support this natively. 
+
+Mutating admission webhooks can be used to partially replicate defaulting for incoming request payloads. But mutating admission webhooks do not run when reading from etcd.
+
+This KEP is about adding support for specifying default values for fields via OpenAPI v3 validation schemas in the CRD manifest. OpenAPI v3 has native support for a `default` field with arbitrary JSON values, for example: 
```

Comment:
will allowing the Default field to be populated trigger any undesireable behavior in the existing go-openapi validator we use? it sounds like we want to implement defaulting as a step separate from validation (which I'm in favor of), but I wanted to check if this will require us to not propagate the specified default to go-openapi when validating

Reply:
the existing validator ignores defaults. and we control what it sees.